src/System.Private.CoreLib/src/System/Environment.Unix.cs

```diff
@@ -71,12 +71,18 @@ public unsafe static String GetEnvironmentVariable(String variable)
             return Encoding.UTF8.GetString((byte*)result, size);
         }
 
-        public static string MachineName
+        private const int MAX_HOST_NAME = 256; // 255 max and null 
+        public static unsafe string MachineName
         {
             get
             {
-                // UNIXTODO: Not yet implemented. Issue: dotnet/corert#650.
-                throw new NotImplementedException();
+                byte *hostName = stackalloc byte[MAX_HOST_NAME];
+                int hostNameLength = Interop.Sys.GetHostName(hostName, MAX_HOST_NAME);
+                if (hostNameLength <= 0)
```

Comment:
I found you set set the host name to empty string. if you do that in the file /etc/hostname, gethostname will return "localhost" name at that time (not empty string). but if you call sethosthost name with empty string then gethostname later will return empty string too. 
I'll fix the code to allow empty string host name